### DISCUSSION TITLE
Could not find valid self-signature in key: Signed digest did not match

### DISCUSSION THREAD

**Discussion by TCFox** [openpgpjs/openpgpjs#1575] [Q&A] [answered]
- OpenPGP.js version: v5.5.0
- Affected platform (Browser or Node.js version): Tested in Chrome and Firefox

I'm having some issues getting GPG keys generated with gpg4win to correctly encrypt a binary (without a signature). It seems at some stage recently, GPG keys that were generated with this software do not appear to work, and I am completely lost as to what might be causing it. To further confuse matters, older keys (generated around about 2021) seem to work perfectly fine, even from the exact same software.

The key was generated today using gpg4win (Kleopatra) with all of the default GUI options (as `raw_pgp_key` in the following Javascript):
```
-----BEGIN PGP PUBLIC KEY BLOCK-----
Comment: User-ID:    test
Comment: Valid from:    20/10/2022 2:17 pm
Comment: Valid until:    20/10/2024 12:00 pm
Comment: Type:    255-bit EdDSA (secret key available)
Comment: Usage:    Signing, Encryption, Certifying User-IDs
Comment: Fingerprint:    66D67EA1B48781F8073E451D1F35DD72304CA4A2

mDMEY1ChkhYJKwYBBAHaRw8BAQdA7B+PqCVJgnvhzz8PiiTqOLXe9u891VtMUOVL
ctrvlIS0BHRlc3SImQQTFgoAQRYhBGbWfqG0h4H4Bz5FHR813XIwTKSiBQJjUKGS
AhsDBQkDw5heBQsJCAcCAiICBhUKCQgLAgQWAgMBAh4HAheAAAoJEB813XIwTKSi
r+0BAOt9evjnCFhCxy3kalDPTX+W47GN8VjOAoD50H1UDnKZAQCHCJXa8PNYC+br
e/Su9cviiTFymfLMuOCMCfPSlKJ6Bbg4BGNQoZISCisGAQQBl1UBBQEBB0CcjevV
vOLBtZqzEU3foqYNGjkjtOaDlJH0u/TL6HzBDQMBCAeIfgQYFgoAJhYhBGbWfqG0
h4H4Bz5FHR813XIwTKSiBQJjUKGSAhsMBQkDw5heAAoJEB813XIwTKSiQUwBALaX
spL21+9NV/RDGFtDyKvDc2ClKXq30zjOYKWqFLkMAQDpRAsgfxSPaJu5qn3aHTPh
cPq0YUyETweK++2FvK0DDA==
=wHJ/
-----END PGP PUBLIC KEY BLOCK-----
```

My code:
```javascript
const openpgp_keys = await Promise.all(raw_pgp_keys.map(armoredKey => openpgp.readKey({ armoredKey })));
const rawFileContents = new Uint8Array(await file.arrayBuffer());
const fileToEncrypt = await openpgp.createMessage({binary: rawFileContents, filename: file.name});
const encryptionOperation = await openpgp.encrypt({
                    message: fileToEncrypt,
                    encryptionKeys: openpgp_keys,
                    format: 'binary',
                    config: { preferredCompressionAlgorithm: openpgp.enums.compression.zip }
});
```

I get the following error (this occurs at the point I run `openpgp.encrypt()`, but appears to be related to when the key is loaded):
```
Uncaught (in promise) Error: Error encrypting message: Could not find valid self-signature in key 1f35dd72304ca4a2: Signed digest did not match
    at fo.verify (signature.js:691:19)
    at async Vo (helper.js:52:63)
    at async pc.getPrimaryUser (key.js:464:15)
    at async helper.js:180:11
    at async Promise.all (index 0)
    at async Xo (helper.js:183:3)
    at async e.encrypt (openpgp.js:285:23)
    at async encryptAndSubmit (mywebpage:99:45)
```

Using this key with desktop software appears to work perfectly fine, the issue looks to be exclusive to OpenPGP.js.

I have no idea what could be causing this, but it feels very much like it could be a bug given this is not a complicated operation, older keys work perfectly fine, and I'm using defaults from popular software (unless I'm missing a `config` parameter I need to supply somewhere!).

Happy to supply further details if needed! I'm rather desperate for an idea as to what might be wrong...

**Comment by TCFox** [comment]
If it helps, I compared the `Key.users[0].selfCertifications[0]` objects between a Windows (gpg4win) generated key and a [macOS](https://gpgtools.org/) generated key that was also made today, and the following properties were notably different:

| Key  | Value (Windows) | Value (macOS) |
|-|-|-|
| hashAlgorithm | 10 | 8 |
| publicKeyAlgorithm | 22 | 1 |
| preferredAEADAlgorithms | [2] | null |
| revoked | null | false |

The Windows-generated key does not work, but the macOS generated key does work.

Just a hunch, could this have something to do with AEAD?

**Comment by larabr** [comment, marked as answer]
Hi @TCFox , I can encrypt fine using the key and code you posted, can you confirm this is the key that gives issues to you?

**Reply by TCFox** [reply]
It's definitely a key that is giving us issues (generated specifically for this example and behaves similarly to issues we've been having with other keys), I'm wondering if something else with my environment is at play here then. I'll dig into it and let you know.

**Reply by larabr** [reply]
I double checked and for me it works both in Chrome and Node.
Maybe check whether `openpgp.config` is being changed somewhere in the code, to e.g. restrict the allowed public key algorithms.

**Comment by larabr** [comment]
I'm converting this to a discussion for now. Feel free to provide more details and we can review whether it is actually an issue on our side :)

**Comment by TCFox** [comment]
After a lot of rather lengthy investigation, it turns out this was due to a very subtle bug in a different library further upstream in our project, which was corrupting our PGP keys in extremely subtle ways. Didn't have anything to do with OpenPGP.js in the end.

Thanks for looking into it though!

**Reply by brianbegy** [reply]
I'm experiencing the same issue calling decrypt in tests from jest.  Can you provide some more detail on the cause?

**Reply by Skelmis** [reply]
We were using https://pypi.org/project/PGPy/ upstream which was then parsing keys from storage to be passed to OpenPGP.js downstream, in the version we use (0.5.4), PGP key content is modified when certain key edge case conditions are encountered without reporting an error so it is hard to notice the subtle difference. This was fixed in a later version.

We only encountered issues with new format(?) keys generated via Kleopatra on windows, other keys appeared to parse just fine.

OpenPGP.js parsed all keys fine when directly supplied, so as far as we are aware there is no problems here and they were only as a result of upstream issues.

cc: @brianbegy
